cpp/ql/src/experimental/Best Practices/GuardedFree.ql

```diff
@@ -0,0 +1,48 @@
+/**
+ * @name Guarded Free
+ * @description NULL-condition guards before function calls to the memory-deallocation
+ *              function free(3) are unnecessary, because passing NULL to free(3) is a no-op.
+ * @kind problem
+ * @problem.severity recommendation
+ * @precision very-high
+ * @id cpp/guarded-free
+ * @tags maintainability
+ *       readability
+ *       experimental
+ */
+
+import cpp
+
+class FreeCall extends FunctionCall {
+  FreeCall() { this.getTarget().hasName("free") }
+}
+
+from IfStmt stmt, FreeCall fc, Variable v
+where
+  stmt.getThen() = fc.getEnclosingStmt() and
+  (
+    stmt.getCondition() = v.getAnAccess() and
+    fc.getArgument(0) = v.getAnAccess()
+    or
+    exists(PointerDereferenceExpr cond, PointerDereferenceExpr arg |
+      fc.getArgument(0) = arg and
+      stmt.getCondition() = cond and
+      cond.getOperand+() = v.getAnAccess() and
+      arg.getOperand+() = v.getAnAccess()
+    )
+    or
+    exists(ArrayExpr cond, ArrayExpr arg |
+      fc.getArgument(0) = arg and
+      stmt.getCondition() = cond and
+      cond.getArrayBase+() = v.getAnAccess() and
+      arg.getArrayBase+() = v.getAnAccess()
+    )
+    or
+    exists(NEExpr eq |
+      fc.getArgument(0) = v.getAnAccess() and
+      stmt.getCondition() = eq and
+      eq.getAnOperand() = v.getAnAccess() and
+      eq.getAnOperand().getValue() = "0"
+    )
+  )
```

Comment:
Have you considered using a `GuardCondition` for this? I think most of this query could be simplified and catch many more things if this code is replaced by:
```ql
from GuardCondition gc, FreeCall fc, Variable v, BasicBlock bb
where
  fc.getArgument(0) = v.getAnAccess() and
  // i.e., `v != 0` in `bb`
  gc.ensuresEq(v.getAnAccess(), 0, bb, false) and
  bb.getANode() = fc
select gc, "unnecessary NULL check before call to $@", fc, "free"
```
(you need to import `semmle.code.cpp.controlflow.Guards` for this to compile).

Reply:
Oh, wait. I should've tested the above before I posted it. There's actually a bug in `GuardCondition` that prevents this from working right now 😨 I've fixed this [here](https://github.com/github/codeql/pull/16333). Once that PR is in, my suggesting does the right thing

Reply:
Thanks! I'll give it a test once it's merged.

@lcartey also pointed me to the [HashCons library](https://codeql.github.com/docs/codeql-language-guides/hash-consing-and-value-numbering/) and that has also simplified the query at seemingly no loss in precision (according to my test):

```QL
from IfStmt stmt, FreeCall fc, Variable v
where
  stmt.getThen() = fc.getEnclosingStmt() and
  (
    hashCons(stmt.getCondition()) = hashCons(fc.getArgument(0))
    or
    exists(NEExpr eq |
      fc.getArgument(0) = v.getAnAccess() and
      stmt.getCondition() = eq and
      eq.getAnOperand() = v.getAnAccess() and
      eq.getAnOperand().getValue() = "0"
    )
  )
select stmt, "unnecessary NULL check before call to $@", fc, "free"
```

Reply:
Indeed, I was also going to suggest that at some point 😂 Basically, there are two libraries that do similar (but not identical things):
- HashCons, which can tell you if two expressions has identical structure and thus will likely evaluate to the same value at runtime.
- [Value numbering](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/valuenumbering/GlobalValueNumbering.qll) (GVN), which can tell you if two expressions will _always_ evaluate to the same value at runtime

For example the arguments to `use` in `use(x); x = x + 1; use(x)` will be equivalent according to HashCons, but they won't be equivalent according to GVN

Reply:
I've tried both HashCons and GVN on my test project [weechat](https://github.com/mario-campos/weechat):

```QL
from IfStmt stmt, FreeCall fc, Variable v
where
  stmt.getThen() = fc.getEnclosingStmt() and
  not globalValueNumber(stmt.getCondition()) = globalValueNumber(fc.getArgument(0)) and
  hashCons(stmt.getCondition()) = hashCons(fc.getArgument(0))
select stmt, "unnecessary NULL check before call to $@", fc, "free"
```

and, interestingly, there is a difference of 10 false negatives between HashCons and GVN&mdash;GVN had the 10 FNs. But, reviewing those FNs, I can't imagine why GVN wouldn't find them; the expressions should be the same at run-time.

https://github.com/mario-campos/weechat/blob/master/src/plugins/trigger/trigger-command.c#L895-L906

In any case, not a major problem; just an interesting note.

Reply:
Unfortunately, that didn't help. It seems that `GuardCondition` does not include cases such as `if (foo)`, only operations:

```QL
from GuardCondition gc, VariableAccess va
where gc.ensuresEq(va, 0, _, false)
select gc, va
```

---

<img width="624" alt="Screenshot 2024-04-29 at 1 04 43 PM" src="https://github.com/github/codeql/assets/1663284/f37062de-5f4f-4b19-ac0d-4a3b4fca9b47">

---

```
$ codeql/codeql version
CodeQL command-line toolchain release 2.17.2+202404261335.
Copyright (C) 2019-2024 GitHub, Inc.
Unpacked in: /Users/mario-campos/codeql
   Analysis results depend critically on separately distributed query and
   extractor modules. To list modules that are visible to the toolchain,
   use 'codeql resolve qlpacks' and 'codeql resolve languages'.
```

Reply:
Sorry about all the back-and-forth. Here's (hopefully) the final fix to make this work: https://github.com/github/codeql/pull/16364 🤞

Reply:
Not a problem. Thank you for the persistence!

Reply:
That's good to know, thanks! We've just merged https://github.com/github/codeql/pull/16364. Do you mind checking if the results are what you expect on your tests _now_? 🙏

Reply:
Okay, I tested it and it's working now! And, I think it found even more cases than I anticipated (1370 vs 1230)!